### PR TITLE
Fix KubernetesPodOperator duplicating logs when interrupted

### LINKED ISSUE #33498: KubernetesPodOperator duplicating logs when they are being interrupted
### Apache Airflow version

main (development)

### What happened

This is a random issue. The KubernetesPodOperator duplicates logs when they are interrupted.

### What you think should happen instead

When the logs are interrupted, the KubernetesPodOperator should continue logging since the last captured timestamp, and not re-print logs that were already printed.

### How to reproduce

If you run the following dag mulltiple times, you might get duplicated logs. Note that the issue is random and not easily reproducible:

```
import datetime

from airflow import models
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator

YESTERDAY = datetime.datetime.now() - datetime.timedelta(days=1)

with models.DAG(
    dag_id="composer_sample_kubernetes_pod",
    schedule_interval=datetime.timedelta(days=1),
    start_date=YESTERDAY,
) as dag:
    timeout = 240
    iterations = 1000
    arguments = \
        'for i in {1..%(iterations)s}; do echo "$i of %(iterations)s"; done' % {'iterations': iterations}

    kubernetes_min_pod_0 = KubernetesPodOperator(
        task_id="pod-ex-minimum-0",
        name="pod-ex-minimum-0",
        cmds=["/bin/bash", "-c"],
        arguments=[arguments],
        namespace="default",
        image="gcr.io/gcp-runtimes/ubuntu_18_0_4",
        startup_timeout_seconds=timeout
    )
```

Here is a sample output of duplicated logs:

[2023-08-18, 13:32:54 UTC] {kublogduplication.py:81} INFO - [base] 994 of 1000
[2023-08-18, 13:32:54 UTC] {kublogduplication.py:81} INFO - [base] 995 of 1000
[2023-08-18, 13:32:54 UTC] {kublogduplication.py:81} INFO - [base] 996 of 1000
[2023-08-18, 13:32:54 UTC] {kublogduplication.py:81} INFO - [base] 997 of 1000
[2023-08-18, 13:32:54 UTC] {kublogduplication.py:81} INFO - [base] 998 of 1000
[2023-08-18, 13:32:54 UTC] {kublogduplication.py:81} INFO - [base] 999 of 1000
[2023-08-18, 13:32:54 UTC] {kublogduplication.py:81} INFO - [base] 1000 of 1000
[2023-08-18, 13:32:54 UTC] {pod_manager.py:450} ERROR - Error parsing timestamp (no timestamp in message ''). Will continue execution but won't update timestamp
[2023-08-18, 13:32:54 UTC] {kublogduplication.py:81} INFO - [base] 
[2023-08-18, 13:32:54 UTC] {kublogduplication.py:116} WARNING - Pod pod-ex-minimum-0-b3bf545v log read interrupted but container base still running
[2023-08-18, 13:32:55 UTC] {before.py:35} INFO - Starting call to 'unusual_prefix_01add01418c7cdaf7afbb9c57d62a05595a85aa7_kublogduplication.CustomPodManager.fetch_container_logs.<locals>.consume_logs', this is the 1st time calling it.
[2023-08-18, 13:32:55 UTC] {kublogduplication.py:81} INFO - [base] 1 of 1000
[2023-08-18, 13:32:55 UTC] {kublogduplication.py:81} INFO - [base] 2 of 1000
[2023-08-18, 13:32:55 UTC] {kublogduplication.py:81} INFO - [base] 3 of 1000
[2023-08-18, 13:32:55 UTC] {kublogduplication.py:81} INFO - [base] 4 of 1000
[2023-08-18, 13:32:55 UTC] {kublogduplication.py:81} INFO - [base] 5 of 1000
[2023-08-18, 13:32:55 UTC] {kublogduplication.py:81} INFO - [base] 6 of 1000
[2023-08-18, 13:32:55 UTC] {kublogduplication.py:81} INFO - [base] 7 of 1000
[2023-08-18, 13:32:55 UTC] {kublogduplication.py:81} INFO - [base] 8 of 1000

### Operating System

Ubuntu

### Versions of Apache Airflow Providers

main

### Deployment

Google Cloud Composer

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by fdemiane**
The KubernetesPodOperator is duplicating logs when they are interrupted. When calling _client.read_namespaced_pod_log, we were passing an outdated since_seconds: we were not using the latest captured timestamp when computing since_seconds.

closes: #33498

### CODE REVIEW

**Comment by eladkal** [comment]
@fdemiane is this PR ready for review?

File: tests/providers/cncf/kubernetes/utils/test_pod_manager.py [hunk 1]
```diff
@@ -252,6 +253,26 @@ def test_parse_log_line(self):
         assert timestamp == pendulum.parse(real_timestamp)
         assert line == log_message
 
+    @mock.patch("airflow.providers.cncf.kubernetes.utils.pod_manager.PodManager.read_pod_logs")
+    def test_fetch_container_logs_returning_last_timestamp(self, mock_read_pod_logs):
+        timestamp_string = "2020-10-08T14:16:17.793417674Z"
+
+        mock_read_pod_logs.return_value = [bytes(f"{timestamp_string} message", "utf-8"), b"notimestamp"]
+
+        self.first_time = True
```
Comment (kosteev) [inline tests/providers/cncf/kubernetes/utils/test_pod_manager.py]: This looks a bit clumsy to define here self variable that is used only in this test.
I can suggest this approach:
```
def mock_container_is_running(...):
    if mock_container_is_running.first_time:
        ...
    ...

mock_container_is_running.first_time = True
```

File: tests/providers/cncf/kubernetes/utils/test_pod_manager.py [hunk 2]
```diff
@@ -252,6 +253,26 @@ def test_parse_log_line(self):
         assert timestamp == pendulum.parse(real_timestamp)
         assert line == log_message
 
+    @mock.patch("airflow.providers.cncf.kubernetes.utils.pod_manager.PodManager.read_pod_logs")
+    def test_fetch_container_logs_returning_last_timestamp(self, mock_read_pod_logs):
+        timestamp_string = "2020-10-08T14:16:17.793417674Z"
+
+        mock_read_pod_logs.return_value = [bytes(f"{timestamp_string} message", "utf-8"), b"notimestamp"]
```
Comment (kosteev) [inline tests/providers/cncf/kubernetes/utils/test_pod_manager.py:264]: Nit: seems that new line above can be omit. Given that there are too many new lines in the method, I would strive to reduce them if possible to make code more logically grouped for better readability.

File: tests/providers/cncf/kubernetes/utils/test_pod_manager.py [hunk 3]
```diff
@@ -252,6 +253,26 @@ def test_parse_log_line(self):
...
+
+        self.first_time = True
+
+        def mock_container_is_running(pod, container_name):
+            if self.first_time:
+                self.first_time = False
+                return True
+            return False
+
+        self.pod_manager.container_is_running = mock_container_is_running
```
Comment (kosteev) [inline tests/providers/cncf/kubernetes/utils/test_pod_manager.py]: This has to be done with mock.patch, otherwise it will be preserved for next tests.

**Review by kosteev — APPROVED** [review]

**Comment by fdemiane** [comment]
> @fdemiane is it ready for review?

@eladkal Yes, I just converted it from a draft PR, thanks.

**Review by amoghrajesh — APPROVED** [review]
This looks pretty good +1

**Review by potiuk — APPROVED** [review]

**Pull request merged**